sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp

```diff
@@ -90,51 +87,48 @@ namespace Azure { namespace Core { namespace Http { namespace _detail {
      * @param connection CURL HTTP connection to add to the pool.
      * @param lastStatusCode The most recent HTTP status code received from the \p connection.
      */
-    static void MoveConnectionBackToPool(
+    void MoveConnectionBackToPool(
         std::unique_ptr<CurlNetworkConnection> connection,
         HttpStatusCode lastStatusCode);
 
-    // Class can't have instances.
-    CurlConnectionPool() = delete;
-
-    static void StopCleaner() { g_isCleanConnectionsRunning = false; }
+    /**
+     * @brief Remove all connections from the pool and reset the connection counter.
+     *
+     */
+    void ResetPool()
+    {
+      ConnectionPoolIndex.clear();
+      ConnectionCounter = 0;
+    }
 
-  private:
     /**
-     * Review all connections in the pool and removes old connections that might be already
-     * expired and closed its connection on server side.
+     * @brief Keeps an unique key for each host and creates a connection pool for each key.
+     *
+     * @details This way getting a connection for a specific host can be done in O(1) instead of
+     * looping a single connection list to find the first connection for the required host.
+     *
+     * @remark There might be multiple connections for each host.
      */
-    static void CleanUp();
+    std::map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>> ConnectionPoolIndex;
 
-    AZ_CORE_DLLEXPORT static uint64_t g_connectionCounter;
-    AZ_CORE_DLLEXPORT static std::atomic<bool> g_isCleanConnectionsRunning;
-    // Removes all connections and indexes
-    static void ClearIndex() { CurlConnectionPool::ConnectionPoolIndex.clear(); }
+    std::mutex ConnectionPoolMutex;
 
-    // Makes possible to know the number of current connections in the connection pool for an
-    // index
-    static int64_t ConnectionsOnPool(std::string const& host)
-    {
-      auto& pool = CurlConnectionPool::ConnectionPoolIndex[host];
-      return pool.size();
-    };
+    std::condition_variable ConditionalVariableForCleanThread;
 
-    // Makes possible to know the number indexes in the pool
-    static int64_t ConnectionsIndexOnPool()
-    {
-      return CurlConnectionPool::ConnectionPoolIndex.size();
-    };
-  };
+    // Keep the count of the total connections from all indexes
+    uint64_t ConnectionCounter;
 
-  struct CurlGlobalStateForAzureSdk
-  {
-    CurlGlobalStateForAzureSdk() { curl_global_init(CURL_GLOBAL_ALL); }
+    AZ_CORE_DLLEXPORT static Azure::Core::Http::_detail::CurlConnectionPool g_curlConnectionPool;
 
-    ~CurlGlobalStateForAzureSdk()
-    {
-      CurlConnectionPool::StopCleaner();
-      curl_global_cleanup();
-    }
+  private:
+    // private constructor to keep this as singleton.
+    CurlConnectionPool() { curl_global_init(CURL_GLOBAL_ALL); }
+
+    std::thread m_cleanThread;
+
+    // Makes possible to know the number of current connections in the connection pool for an
+    // index
```

Comment:
For a given index? Don't you mean host?
```suggestion
    // Makes possible to know the number of current connections in the connection pool for a
    // given host.
```

Reply:
index, key, host, connection-key.  There's no official name hehehe

Reply:
The parameter name to this function is `host`, so I would at least match that.

Otherwise, rename it to `connection_key` and refer to it in the comment as well.